<a href="https://colab.research.google.com/github/lcbjrrr/algojust/blob/main/Pmpt04_hist_algo_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algorithmic and Historical bias



- **Algorithmic bias**: Bias that arises from the design, implementation, or use of algorithms. This can be due to skewed training data or poor algorithm design and can lead to unfair outcomes in areas like recruiting or criminal justice.
Omitted variable bias: Occurs when important variables are left out of a model or analysis, leading to inaccurate results. For example, analyzing car data without considering mileage or age can lead to inaccurate conclusions about vehicle value

- **Historical bias**: Occurs when past socio-cultural prejudices and beliefs are reflected in the data and subsequent analysis. This is particularly challenging when historical data is used to train machine learning models, as the models can perpetuate and amplify these biases. An example is an AI hiring tool that learned from historically biased hiring data and ended up favoring male candidates
*The data used herein is for illustrative purposes only and does not reflect actual real-world data*.



#### Here is a trained [model](https://colab.research.google.com/github/lcbjrrr/quantai/blob/main/M2_Py_ML_RegLin.ipynb), please use it to perform preditions.



## AI/ML Models



Machine Learning (ML) models are a technology where computers learn from data to find patterns and make predictions without needing explicit programming for each task. These models improve their accuracy by processing more data and adjusting how they analyze information over time.


Donwload: [car_insurance-bias.csv](https://raw.githubusercontent.com/lcbjrrr/algojust/refs/heads/main/car_insurance%20-%20bias.csv)


> Read the file 'car_insurance - bias.csv' and show the dataframe

In [1]:
import pandas as pd

df = pd.read_csv('/content/car_insurance - bias.csv')
df

,age,female,premium,vehicle_age,claim_last2y,v
0,23,1,580,3,0,Sedan
1,24,0,520,5,1,SUV
2,31,1,510,2,0,Hatchback
3,33,0,450,7,1,Sedan
4,40,1,440,4,0,SUV
5,44,0,380,6,1,Truck
6,55,1,350,3,0,Sedan
7,56,0,310,8,1,SUV
8,18,1,650,1,0,Sedan
9,19,0,590,2,1,Hatchback


## Training the model



The get_dummies function in pandas is used for one-hot encoding categorical variables. It converts categorical data into a format that can be easily understood by machine learning algorithms, where each category becomes a new column with binary values (0 or 1).



> Transform the categorical variable into numerical ones

In [2]:
df = pd.get_dummies(df, columns=['v'], prefix='v')
df

,age,female,premium,vehicle_age,claim_last2y,v_Hatchback,v_SUV,v_Sedan,v_Truck
0,23,1,580,3,0,False,False,True,False
1,24,0,520,5,1,False,True,False,False
2,31,1,510,2,0,True,False,False,False
3,33,0,450,7,1,False,False,True,False
4,40,1,440,4,0,False,True,False,False
5,44,0,380,6,1,False,False,False,True
6,55,1,350,3,0,False,False,True,False
7,56,0,310,8,1,False,True,False,False
8,18,1,650,1,0,False,False,True,False
9,19,0,590,2,1,True,False,False,False


To train a AI/ML model, in the example here a linear regression model using a training dataset, save the trained model to a file, load it back, and then use it to make predictions. The process involves importing libraries, fitting the model to data, evaluating its performance, and applying it to new data points to see the predicted outcomes. It is possible to measure the correctness in traning with the score() method, in the case here (RegLing) the r2.



> Train a linear regression model, so it learn how to predict Premium based on age, female, vehicle_age, claim_last2y, vehicle type. Then show the score (R2) of the training.

In [3]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# Define features (X) and target (y)
X = df[['age', 'female', 'vehicle_age', 'claim_last2y', 'v_Hatchback', 'v_SUV', 'v_Sedan', 'v_Truck']]
y = df['premium']

# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X, y)

# Calculate and print the R2 score
r2_score = model.score(X, y)
print(f"R2 Score: {r2_score:.2f}")

R2 Score: 0.98


> Create a test dataset with 3 persons with a hatchback,
one being a man and the other two women,
The first two have a claim in the last two years,
and the other has no. All three with a 2-year-old car.
Add another 33-year-old woman, but with a 4-year-old car.
Then add two people of 66, one man and the other a woman,
both with a 4-year-old SUV, and with no claim in the last 2 years

In [5]:
import pandas as pd

test_data = [
    # 3 persons with a hatchback, one man, two women. First two claim, last one no claim. All 2-year-old car.
    {'age': 30, 'female': 0, 'vehicle_age': 2, 'claim_last2y': 1, 'v_Hatchback': True, 'v_SUV': False, 'v_Sedan': False, 'v_Truck': False},
    {'age': 35, 'female': 1, 'vehicle_age': 2, 'claim_last2y': 1, 'v_Hatchback': True, 'v_SUV': False, 'v_Sedan': False, 'v_Truck': False},
    {'age': 40, 'female': 1, 'vehicle_age': 2, 'claim_last2y': 0, 'v_Hatchback': True, 'v_SUV': False, 'v_Sedan': False, 'v_Truck': False},
    # Another 33-year-old woman, but with a 4-year-old car (assuming Sedan for diversification, no claim).
    {'age': 33, 'female': 1, 'vehicle_age': 4, 'claim_last2y': 0, 'v_Hatchback': False, 'v_SUV': False, 'v_Sedan': True, 'v_Truck': False},
    # Two people of 66, one man and the other a woman, both with a 4-year-old SUV, and with no claim in the last 2 years.
    {'age': 66, 'female': 0, 'vehicle_age': 4, 'claim_last2y': 0, 'v_Hatchback': False, 'v_SUV': True, 'v_Sedan': False, 'v_Truck': False},
    {'age': 66, 'female': 1, 'vehicle_age': 4, 'claim_last2y': 0, 'v_Hatchback': False, 'v_SUV': True, 'v_Sedan': False, 'v_Truck': False}
]

# Create DataFrame
X_test = pd.DataFrame(test_data)

# Ensure column order matches the training features (X from previous cell)
X_test = X_test[['age', 'female', 'vehicle_age', 'claim_last2y', 'v_Hatchback', 'v_SUV', 'v_Sedan', 'v_Truck']]

print("Test Dataset (X_test):")
X_test

Test Dataset (X_test):


,age,female,vehicle_age,claim_last2y,v_Hatchback,v_SUV,v_Sedan,v_Truck
0,30,0,2,1,True,False,False,False
1,35,1,2,1,True,False,False,False
2,40,1,2,0,True,False,False,False
3,33,1,4,0,False,False,True,False
4,66,0,4,0,False,True,False,False
5,66,1,4,0,False,True,False,False


> Create predictions on this dataset and save it as a column on the test dataset.


In [6]:
X_test['predicted_premium'] = model.predict(X_test)
X_test

,age,female,vehicle_age,claim_last2y,v_Hatchback,v_SUV,v_Sedan,v_Truck,predicted_premium
0,30,0,2,1,True,False,False,False,477.201246
1,35,1,2,1,True,False,False,False,480.469855
2,40,1,2,0,True,False,False,False,445.338214
3,33,1,4,0,False,False,True,False,496.818075
4,66,0,4,0,False,True,False,False,217.513302
5,66,1,4,0,False,True,False,False,256.802817


*From the test dataset clearly indicate that for individuals with otherwise identical risk profiles (age, vehicle age, claim history, vehicle type), female individuals are consistently predicted to pay higher insurance premiums than male individuals. For example: 33-year-olds with a 2-year-old hatchback and a recent claim: A male is predicted to pay 455.59, while a female is predicted to pay 494.88. 66-year-olds with a 4-year-old SUV and no recent claim: A male is predicted to pay 217.51, while a female is predicted to pay 256.80.
This pattern suggests that the model is discriminating based on gender, which is a form of bias. This could stem from historical biases present in the training data (if historical premiums were higher for women without justified actuarial reasons) or from an algorithmic design that inadvertently amplifies gender as a predictive factor even when it shouldn't be*



## Linear Regression 101

Dowload: [grades-all.csv](https://raw.githubusercontent.com/lcbjrrr/algojust/refs/heads/main/grades%20-%20all.csv)



> Read the file grades-all.csv and create a linear regression model that predicts Ex3 based on Ex1 and Ex2.



In [7]:
import pandas as pd
from sklearn.linear_model import LinearRegression

# Read the CSV file
grades_df = pd.read_csv('/content/grades - all.csv')

# Define features (X) and target (y)
X = grades_df[['Ex1', 'Ex2']]
y = grades_df['Ex3']

# Initialize and train the Linear Regression model
model_grades = LinearRegression()
model_grades.fit(X, y)

# Calculate and print the R2 score
r2_score_grades = model_grades.score(X, y)
print(f"R2 Score for grades model: {r2_score_grades:.2f}")

# Display the first few rows of the DataFrame to confirm loading
print("\nFirst 5 rows of grades_df:")
print(grades_df.head())

R2 Score for grades model: 0.98

First 5 rows of grades_df:
   Ex1  Ex2  Ex3
0  100  100   90
1   90  100   90
2   95  100  100
3   94   99   99
4   93   98   98




> Predict Ex3 for this test dataset: (80, 80); (20, 40); (35, 40)



In [10]:
import pandas as pd

# Create a DataFrame for the test dataset
test_data_grades = pd.DataFrame({
    'Ex1': [80, 20, 35],
    'Ex2': [80, 40, 40]
})
print(test_data_grades)
# Make predictions using the trained model_grades
model_grades.predict(test_data_grades)

   Ex1  Ex2
0   80   80
1   20   40
2   35   40


array([73.12693817, 39.75624675, 34.2650204 ])

## Activity: Algorithmic and Historical bias




**Problem**

Even foundational models like linear regression can perpetuate and amplify biases present in their training data, leading to unfair or inaccurate predictions. Your task is to select a dataset from Kaggle suitable for linear regression. Train a linear regression model on your chosen dataset and generate predictions. Then analyze your predictions for evidence of bias (in case there are).

**Conclusions**

Summarize the specific biases identified in your model's predictions, detailing their impact. Discuss the effectiveness of any mitigation strategies you attempted or propose and reflect on the ethical implications of your findings. Your conclusion should articulate the challenges of achieving fairness in machine learning and suggest future research directions or practical steps for developing more equitable predictive models. In case there are no signs of bias, provide concrete evidence of it and comment on this finding.

## Appendix



```
!wget https://github.com/lcbjrrr/algojust/raw/refs/heads/main/lingreg.pkl
```

```
import joblib
model = joblib.load('lingreg.pkl')
model.coef_
```

The joblib.load function in Python quickly loads saved data like machine learning models, especially when working with large arrays. .pkl files are Python-specific files created with the pickle module to store objects in a serialized format for later use.​

With the trained model, we can perform predictions. Predictions are estimates or decisions made by trained models when given new, unseen data, using patterns learned from historical information.​

```
joblib.dump(linreg, 'lingreg.pkl')
```